File: programmers/kakao#2/java/KAKAO6.java
import java.util.*;

public class KAKAO6 {
    class Solution {
        public int solution(String dartResult) {
            int answer = 0;
            
            List<Integer> nums = new ArrayList<>();
            
            int prev = -1;
            for (int i=0;i<dartResult.length();i++) {
                if (!Character.isDigit(dartResult.charAt(i))) {
                    if (dartResult.substring(prev+1, i).length() != 0) {
                        nums.add(Integer.parseInt(dartResult.substring(prev+1, i)));
                    }
                    prev = i;
                    
                    char oper = dartResult.charAt(i);
                    
                    if (oper == 'D') {
                        nums.set(nums.size()-1, (int)Math.pow(nums.get(nums.size()-1), 2));
                    } else if (oper == 'T') {
                        nums.set(nums.size()-1, (int)Math.pow(nums.get(nums.size()-1), 3));
                    } else if (oper == '*') {
                        nums.set(nums.size()-1, nums.get(nums.size()-1)*2);
                        if (nums.size() > 1)
                            nums.set(nums.size()-2, nums.get(nums.size()-2)*2);
                    } else if (oper == '#') {
                        nums.set(nums.size()-1, nums.get(nums.size()-1)*(-1));
                    }
                }
            }
            
            for (int num : nums) {
                answer += num;
            }
            
            return answer;
        }
    }
}

/**
 * [카카오 2018 공채] 다트 게임
 * 해결: O
 * 시간: 22분
 */

File: programmers/kakao#2/python/KAKAO6.py
def solution(dartResult):
    stack = []
    sdt = {"S": 1, "D": 2, "T": 3}
    for i in range(len(dartResult)):
        if not dartResult[i].isdigit():
            # 기호가 발견되면 인덱스 내려가며 숫자 찾기
            idx = i-1
            while idx >= 0 and dartResult[idx].isdigit():
                idx -= 1
            # 숫자 발견한 경우 스택에 추가
            if dartResult[idx+1:i].isdigit():
                stack.append(int(dartResult[idx+1:i]))

            if dartResult[i] in sdt:
                if stack:
                    stack[-1] = stack[-1] ** sdt[dartResult[i]]
            elif dartResult[i] == "*":
                if stack:
                    stack[-1] *= 2
                    if len(stack) >= 2:
                        stack[-2] *= 2
            elif dartResult[i] == "#":
                if stack:
                    stack[-1] *= -1
        
    return sum(stack)


"""
[카카오 2018 공채] 다트 게임
해결: O
시간: 22분
"""